kore/test/Test/Kore/Step/Axiom/Matcher.hs

```diff
@@ -1029,6 +1040,195 @@ test_unificationWithAppMatchOnTop =
+matchingList =
+    [ testCase "concrete top" $ do
+        let expect = top
+        actual <- matchConcrete [1, 2] [1, 2]
+        assertEqualWithExplanation "" expect actual
+    , testCase "concrete bottom" $ do
+        let expect = Nothing
+        actual <- matchConcrete [1, 2] [1, 3]
+        assertEqualWithExplanation "" expect actual
+    , testCase "single variable" $ do
+        let expect = substitution [(Mock.xInt, 2)]
+        actual <- matchVariable [Right 1, Left Mock.xInt] [1, 2]
+        assertEqualWithExplanation "" expect actual
+    , testCase "two variables" $ do
+        let expect = substitution [(Mock.xInt, 2), (Mock.yInt, 4)]
+        actual <- matchVariable
+            [ Right 1
+            , Left Mock.xInt
+            , Right 3
+            , Left Mock.yInt
+            ]
+            [ 1, 2, 3, 4 ]
+        assertEqualWithExplanation "" expect actual
+    , testCase "no AC" $ do
+        let expect = Nothing
+        actual <- matchVariable
+            [ Right 1, Left Mock.xInt] [ 2, 1 ]
+        assertEqualWithExplanation "" expect actual
+    ]
+  where
+    top = Just $ OrPredicate.fromPredicate Predicate.topPredicate
+    substitution subst = Just $ MultiOr.make
+        [ Conditional
+            { term = ()
+            , predicate = makeTruePredicate
+            , substitution =
+                Substitution.unsafeWrap
+                    ((fmap . fmap) mkInt subst)
+            }
+        ]
+    mkInt = Int.asInternal Mock.intSort
+    matchList =
+        matchDefinition `on`
+            (List.asInternal Mock.metadataTools Mock.listSort . Seq.fromList)
+    matchConcrete =
+        matchList `on` fmap mkInt
+    matchVariable var val =
+            matchList
+            (either mkVar mkInt <$> var)
+            (mkInt <$> val)
+
+matchingMap :: [TestTree]
+matchingMap =
+    [ testCase "concrete top" $ do
+        let expect = top
+        actual <- matchConcrete [(1, 10), (2, 11)] [(1, 10), (2, 11)]
+        assertEqualWithExplanation "" expect actual
+    , testCase "concrete bottom" $ do
+        let expect = Nothing
+        actual <- matchConcrete [(1, 10), (2, 11)] [(1, 10)]
+        assertEqualWithExplanation "" expect actual
+    , testCase "single variable" $ do
+        let expect = substitution [(Mock.xInt, 11)]
+        actual <- matchVariable
+            [ (1, Right 10)
+            , (2, Left Mock.xInt)
+            ]
+            [ (2, 11)
+            , (1, 10)
+            ]
+        assertEqualWithExplanation "" expect actual
+    , testCase "two variables" $ do
+        let expect = substitution [(Mock.xInt, 11), (Mock.yInt, 13)]
+        actual <- matchVariable
+            [ (1, Right 10)
+            , (2, Left Mock.xInt)
+            , (3, Right 12)
+            , (4, Left Mock.yInt)
+            ]
+            [ (2, 11)
+            , (1, 10)
+            , (4, 13)
+            , (3, 12)
+            ]
+        assertEqualWithExplanation "" expect actual
```

Comment:
We should be able to handle the case `match [ Mock.xMap , ( 1 , 2 ) ] [ (1, 2) , (3, 4), (5, 6) ]` which should get substitution `Mock.xMap -> [ (3, 4), (5, 6) ]`.